subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperChecksumVerificationTest.groovy

```diff
@@ -72,4 +75,53 @@ Expected checksum: 'bad'
         then:
         success.output.contains('BUILD SUCCESSFUL')
     }
+
+    def "wrapper requires checksum configuration if a checksum is present in gradle-wrapper.properties"() {
+        given:
+        prepareWrapper(new URI("${server.uri}/gradle-bin.zip"))
+
+        and:
+        file('gradle/wrapper/gradle-wrapper.properties') << "distributionSha256Sum=${Hashing.sha256().hashFile(distribution.binDistribution).toZeroPaddedString(Hashing.sha256().hexDigits)}"
+
+        when:
+        def result = wrapperExecuter.withTasks("wrapper", "--gradle-version", "7.5").runWithFailure()
+
+        then:
+        result.assertHasErrorOutput("gradle-wrapper.properties contains distributionSha256Sum property, but the wrapper configuration does not have one. Specify one in the wrapped task configuration or with the --gradle-distribution-sha256-sum task option")
+    }
+
+    def "wrapper uses new checksum if it was provided as an option"() {
+        given:
+        prepareWrapper(new URI("${server.uri}/gradle-bin.zip"))
+
+        and:
+        file('gradle/wrapper/gradle-wrapper.properties') << "distributionSha256Sum=${Hashing.sha256().hashFile(distribution.binDistribution).toZeroPaddedString(Hashing.sha256().hexDigits)}"
+
+        when: "Run wrapper to update with released distribution checksum and url"
+        def releasedDistribution = IntegrationTestBuildContext.INSTANCE.distribution("7.5")
+        def releasedDistributionUrl = releasedDistribution.binDistribution.toURI().toString()
+        def releasedDistributionChecksum = Hashing.sha256().hashFile(releasedDistribution.binDistribution).toZeroPaddedString(Hashing.sha256().hexDigits)
+        wrapperExecuter.withTasks("wrapper", "--gradle-distribution-url", releasedDistributionUrl, "--gradle-distribution-sha256-sum", releasedDistributionChecksum).run()
```

Comment:
I've spend a couple hours but still don't understand how to properly setup my test environment to prevent downloading of another distribution of Gradle.

Reply:
I was thinking about something like this:
```
    def "wrapper uses new checksum if it was provided as an option"() {
        given:
        prepareWrapper(new URI("${server.uri}/gradle-bin.zip"))

        and:
        file('gradle/wrapper/gradle-wrapper.properties') << "distributionSha256Sum=${Hashing.sha256().hashFile(distribution.binDistribution).toZeroPaddedString(Hashing.sha256().hexDigits)}"

        when: "Run wrapper to update with released distribution checksum and url"
        def releasedDistributionUrl = "https://example.com/gradle-0.0-bin.zip"
        def releasedDistributionChecksum = "abcdefghijklmnopqrstuvwxyz"
        wrapperExecuter.withTasks("wrapper", "--gradle-distribution-url", releasedDistributionUrl, "--gradle-distribution-sha256-sum", releasedDistributionChecksum).run()

        then:
        file('gradle/wrapper/gradle-wrapper.properties').getProperties().get(WrapperExecutor.DISTRIBUTION_SHA_256_SUM) == releasedDistributionChecksum
    }
```

It doesn't matter if the URL is invalid because we'll never try to download it and check the checksum.